### PR TITLE
chore(CPlusPlus): adding recursion programs

### PR DESCRIPTION
**Pull request by jishanshaikh4**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] New algorithm
- [ ] Optimization in previous algorithms
- [ ] Code style update
- [ ] Refactor
- [x] Documentation
- [ ] Other, please describe:

**Briefly describe the changes in this PR:**

Added following programs to the section ```algorithms/cpp/recursion```:
- GCD/HCF of two numbers
- Sum of all elements of an array
- Decimal number to Binary conversion
- Sum of digits of a decimal integer

**Other information:** Refer #395 and #347

### CODE REVIEW

**Review by Arsenic-ATG — CHANGES_REQUESTED** [review]
Thanks for the executable links with the programs.

I have left the review comments in the source itself ( some of them are nit picks ).

**Comment by jishanshaikh4** [comment]
@Arsenic-ATG Thank you for the review. Make appropriate changes as you like. It is allowed for maintainers.

**Comment by Arsenic-ATG** [comment]
> @Arsenic-ATG Thank you for the review. Make appropriate changes as you like. It is allowed for maintainers.

If it would have been small nits or doc updates, I would have done that easily.
But some stuff here requires the creator's involvement ( especially decimal-too-binary.cpp ) where the change might involve refactoring the way program works altogether.

**Comment by ming-tsai** [comment]
@jishanshaikh4, please solve the merge conflict

**Comment by juhi1999** [comment]
Hello I'm new to contribution.. and I wish to contribute codes...

**Comment by Arsenic-ATG** [comment]
> Hello I'm new to contribution.. and I wish to contribute codes...

Greetings @juhi1999, 
In order to contribute, either pick up an [open issue](https://github.com/MakeContributions/DSA/issues) and start working on it, or directly send a PR implementing an algorithm or data-structure in a particular programming language, that is still not present in the repository. 

Also, do make sure to give the repo's [contribution guidelines](https://github.com/MakeContributions/DSA#contribution-guidelines) a good read, before making any contribution.

--- 

P.S. : 
Avoid sending messages in a pull request ( especially if the message is not related to the PR ), such messages create noise in the conversation. 
Kindly open a new issue or use [Discussions](https://github.com/MakeContributions/DSA/discussions) section for such queries.

**Comment by jishanshaikh4** [comment]
@ming-tsai Sorry for the untimely reply, I'm out of my place for a couple of weeks. I would resume as I get back.

**Comment by ming-tsai** [comment]
> @ming-tsai Sorry for the untimely reply, I'm out of my place for a couple of weeks. I would resume as I get back.

Don't worries

**Comment by ming-tsai** [comment]
Hi @jishanshaikh4,
Any update?

**Comment by jishanshaikh4** [comment]
> Hi @jishanshaikh4,
> Any update?

Yep, I'm all on it now.

File: algorithms/CPlusPlus/Recursion/decimal-to-binary-conversion.cpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// Program to convert Decimal to Binary
+// Executable link: https://ide.geeksforgeeks.org/8oiG0di9QE
+
+#include <iostream>
+
+using namespace std;
+
+// Trap: Remember the limit of int in C++
+// if a binary of a number exceed the limit of int, it will overflow
+// Tip: Use long long (and unsigned long long) instead
```
[the remark this answers is not in the document]
Comment (jishanshaikh4) [reply algorithms/CPlusPlus/Recursion/decimal-to-binary-conversion.cpp:10]: Sure thing. I added the comment to provide a bit for beginners.

File: algorithms/CPlusPlus/Recursion/decimal-to-binary-conversion.cpp [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+unsigned long long decimalToBinary (int decimal_number) {
+    if (decimal_number == 0)
+        return 0;
+    else
+        return (decimal_number%2 + 10*decimalToBinary(decimal_number/2));
```
[the remark this answers is not in the document]
Comment (jishanshaikh4) [reply algorithms/CPlusPlus/Recursion/decimal-to-binary-conversion.cpp:16]: Bug fixed.

File: algorithms/CPlusPlus/Recursion/GCD-of-two-numbers.cpp [hunk 3]
```diff
@@ -0,0 +1,20 @@
+// Program to find GCD of two numbers using recursion
+// Executable link: https://ide.geeksforgeeks.org/0ie6920wBA
+
+#include <iostream>
+
+using namespace std;
+
+// Also known as Euclidean algorithm
+// There exists an inbuilt function for gcd also __gcd(T a, T b) with Template T
```
[the remark this answers is not in the document]
Comment (jishanshaikh4) [reply algorithms/CPlusPlus/Recursion/GCD-of-two-numbers.cpp:9]: __gcd() still works in C++

File: algorithms/CPlusPlus/Recursion/GCD-of-two-numbers.cpp [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+int gcd(int a, int b) {
+    if (b == 0)
+        return a;
+    return gcd(b, a%b);
+}
+
+int main() {
+    int n1 = 56, n2 = 65;
+    cout << "GCD of " << n1 << " and " << n2 << " is " << gcd(n1, n2);
```
[the remark this answers is not in the document]
Comment (jishanshaikh4) [reply algorithms/CPlusPlus/Recursion/GCD-of-two-numbers.cpp]: Sure thing. gcd() has been replaced to euclidean_gcd().

**Comment by jishanshaikh4** [comment]
@Arsenic-ATG Please review again.

**Review by Arsenic-ATG — APPROVED** [review]

**Comment by ming-tsai** [comment]
Nice work!!!

**Pull request merged**